### PR TITLE
#6: Integration Test Home Route

### PR DESCRIPTION
**Pull request by shmbajaj**
## Summary

Integration tests for HomePage route/component.
The tests are written from user POV, How user will behave and interact with home page.

`NOTE:` To run tests local-backend is expected to be running

### 1. Renders Home Page:
- **Objective:** Verify the correct rendering of the home page.
- **Actions:** Render `HomePage` within a `BrowserRouter`.
- **Assertions:** Confirm the presence of key elements like "WanderLust" text, "Create post" button, and loading skeletons.

### 2. Calls Mocked `useNavigate` Function:
- **Objective:** Ensure the "Create post" button triggers the `useNavigate` function.
- **Actions:** Render `HomePage` and click the "Create post" button.
- **Assertions:** Verify the `mockedUseNavigate` function is called once.

### 3. Renders Home Page with BlogFeed:
- **Objective:** Confirm correct rendering of the BlogFeed section.
- **Actions:** Render `HomePage`, click a category pill, and check for related posts.
- **Assertions:** Validate the BlogFeed rendering and category pill functionality.

### 4. Navigates on Featured Post Click:
- **Objective:** Check if clicking a featured post card navigates to the details page.
- **Actions:** Render `HomePage` and click the first featured post card.
- **Assertions:** Ensure the `mockedUseNavigate` function is called once.

### 5. Renders Home Page with All Posts:
- **Objective:** Verify the correct rendering of all post cards.
- **Actions:** Render `HomePage`.
- **Assertions:** Confirm the expected number of post cards are rendered.

### 6. Navigates on All-Post Click:
- **Objective:** Check if clicking on an individual post card navigates to the details page.
- **Actions:** Render `HomePage` and click the first post card.
- **Assertions:** Ensure the `mockedUseNavigate` function is called once.

These tests cover various aspects of the `HomePage` component, including rendering, user interactions, and navigation behavior. The use of testing libraries facilitates thorough testing of the React application.

- No Tests for api **Failed to display home page with blog-feed/all-posts** because firstly shouldn't mock api for integration tests, and second tests are written with user-pov and third user is not ack with failed api responses.

## Miscallenous FEYI 
Cannot mock backend for testcases because using jest, vite and msw results into following errors, which can result into series of errors:
      - [Request/Response/TextEncoder is not defined (Jest)](https://mswjs.io/docs/migrations/1.x-to-2.x/#requestresponsetextencoder-is-not-defined-jest)
      - [Cannot find module ‘msw/node’ (JSDOM)](https://mswjs.io/docs/migrations/1.x-to-2.x/#cannot-find-module-mswnode-jsdom)
      - TS Error: Definitions of the following identifiers conflict with those in another file.

## Issue(s) Addressed
 
#6

## Images
Screenshot 2024-01-06 at 12 17 33 AM

## Prerequisites

- [X] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Comment by shmbajaj** [comment]
> A unit test is not a unit test if it uses these external dependencies - that would rather be an integration test.

**Comment by krishnaacharyaa** [comment]
Hi @shmbajaj, Big thanks from bottom of my heart for this wonderful efforts and timely updates and fast paced PR, This PR contains quite a few updates, can we arrange a call when we both and @chinmaykunkikar is free and review this PR and merge? 

p.s: It would be great if you could add snapshots of the test cases, so that i can get a overview...

**Comment by shmbajaj** [comment]
> Hi @shmbajaj, Big thanks from bottom of my heart for this wonderful efforts and timely updates and fast paced PR, This PR contains quite a few updates, can we arrange a call when we both and @chinmaykunkikar is free and review this PR and merge?
> 
> p.s: It would be great if you could add snapshots of the test cases, so that i can get a overview...

Firstly Thanks @krishnaacharyaa.

Honestly I didn't understood what do you mean by `snapshots`.
Also we can connect today at 09:00PM/2100HRS.

**Comment by krishnaacharyaa** [comment]
> Honestly I didn't understood what do you mean by snapshots.

I got it in the description :)

**Comment by krishnaacharyaa** [comment]
> Calls Mocked useNavigate Function:

Can we replace this with if the add Blog page is visible? So that it will be uniform, the title doesn't seem very promising for a test case in my perspective...

And just another thing, I am still not satisfied with the naming to be honest, dont be mad at me, i am a bit particular....
Presently it is 
 Integration Test: Home Route
   1. Integration Test: renders home page
   2. Integration Test: calls the mockedUseNavigate function
   3. Integration Test: renders home page with BlogFeed
   4. ...

Can we have it as
 Integration Test: Home Route
   1. Home Route: Renders Home Page
   2. Home Route: Renders Add Page
   3. Home Route: Renders posts in BlogFeed Component
   4. ...
   
   I hope even you appreciate the symmetry and easy to understand yet simple titles, which gives you the idea in the first look

**Comment by krishnaacharyaa** [comment]
Update:
This PR is hold because for some reason, i am not convinced with this test case: "Home Route: calls the mockedUseNavigate function"
-> We can either remove this or replace with "Home Route: Renders Add Page on button click"
If you feel this actually makes sense, kindly explain it me, once. If convinced lets merge this....

File: frontend/src/__tests__/integration/home-page.test.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+jest.mock('react-router-dom', () => ({
+  ...(jest.requireActual('react-router-dom') as Record<string, any>),
+  useNavigate: () => mockedUseNavigate,
+}));
+
+afterEach(() => mockedUseNavigate.mockRestore());
+
+describe('Integration Test: Home Route', () => {
+  test('Home Route: renders home page', async () => {
+    //ARRANGE
```
Comment (krishnaacharyaa) [inline frontend/src/__tests__/integration/home-page.test.tsx:17]: Can we rename this to `Home Route: Renders home page`, Caps after colon.... Just to follow the some kind of uniformity, In every test case title

File: frontend/src/__tests__/integration/home-page.test.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      })
+    ).toBeInTheDocument();
+    expect(screen.getByText(/Featured Posts/)).toBeInTheDocument();
+    expect(screen.getAllByTestId('featurepostcardskeleton')).toHaveLength(5);
+    expect(screen.getAllByTestId('latestpostcardskeleton')).toHaveLength(5);
+    expect(screen.getByText(/All Posts/)).toBeInTheDocument();
+    expect(screen.getAllByTestId('postcardskeleton')).toHaveLength(8);
+  });
+  test('Home Route: calls the mockedUseNavigate function', async () => {
+    //ARRANGE
```
Comment (krishnaacharyaa) [inline frontend/src/__tests__/integration/home-page.test.tsx:40]: Home Route: Verify navigation on create post button click

File: frontend/src/__tests__/integration/home-page.test.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    expect(mockedUseNavigate).toHaveBeenCalledTimes(0);
+
+    //ACT
+    await userEvent.click(createPost);
+
+    //ASSERT
+    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
+  });
+  test('Home Route: renders home page with BlogFeed', async () => {
+    //ARRANGE
```
Comment (krishnaacharyaa) [inline frontend/src/__tests__/integration/home-page.test.tsx:60]: Home Route: Verify filtered posts render on category button click

File: frontend/src/__tests__/integration/home-page.test.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      name: 'Nature',
+    });
+    expect(natureCategoryPill).toBeInTheDocument();
+    await userEvent.click(natureCategoryPill);
+    expect(await screen.findByText('Posts related to "Nature"')).toBeInTheDocument();
+    // Strange test got passed api response is 3 over local backend
+    expect(await screen.findAllByTestId('featuredPostCard')).toHaveLength(5);
+  });
+  test('Home Route: on featured post click navigates to /details-page/:title/:id page', async () => {
+    //ARRANGE
```
Comment (krishnaacharyaa) [inline frontend/src/__tests__/integration/home-page.test.tsx:83]: Home Route: Verify navigation on post card click under Featured Posts section

File: frontend/src/__tests__/integration/home-page.test.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    );
+    //ACT
+    //ASSERT
+    const featuredPostCard = await screen.findAllByTestId('featuredPostCard');
+    expect(featuredPostCard).toHaveLength(5);
+    await userEvent.click(featuredPostCard[0]);
+    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
+  });
+  test('renders home page with all post', async () => {
+    //ARRANGE
```
Comment (krishnaacharyaa) [inline frontend/src/__tests__/integration/home-page.test.tsx:97]: Home Route: Verify render of post card under All Post section

File: frontend/src/__tests__/integration/home-page.test.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        <HomePage />
+      </BrowserRouter>
+    );
+    //ACT
+    //ASSERT
+    expect(await screen.findAllByTestId('postcard')).toHaveLength(10);
+  });
+  test('Home Route: on all-post post click navigates to /details-page/:title/:id page', async () => {
+    //ARRANGE
+    render(
```
Comment (krishnaacharyaa) [inline frontend/src/__tests__/integration/home-page.test.tsx:109]: Home Route: Verify navigation on post card click under All Posts section

**Review by krishnaacharyaa — COMMENTED** [review]
Please replace the test case titles with this and then we are good to merge this

**Comment by krishnaacharyaa** [comment]
@shmbajaj, can we close this today

**Comment by shmbajaj** [comment]
@krishnaacharyaa closing this PR because of hells of commits created.

**Pull request closed without merging**